libraries/hotstuff/qc_chain.cpp

```diff
@@ -0,0 +1,1224 @@
+#endif
+
+      //update internal state
+      update(proposal);
+
+      for (auto &msg : msgs) {
+         send_hs_vote_msg(msg);
+      }
+
+      //check for leader change
+      leader_rotation_check();
+
+      //auto total_time = fc::time_point::now() - start;
+      //if (_log) ilog(" ... process_proposal() total time : ${total_time}", ("total_time", total_time));
+   }
+
+   void qc_chain::process_vote(const hs_vote_message & vote){
+
+      //auto start = fc::time_point::now();
+#warning check for duplicate or invalid vote. We will return in either case, but keep proposals for evidence of double signing
+      //TODO: check for duplicate or invalid vote. We will return in either case, but keep proposals for evidence of double signing
+
+      bool am_leader = am_i_leader();
+
+      if (!am_leader)
+         return;
+#ifdef QC_CHAIN_TRACE_DEBUG
+      ilog(" === Process vote from ${finalizer} : current bitset ${value}" , ("finalizer", vote.finalizer)("value", _current_qc.active_finalizers));
+#endif
+      // only leader need to take action on votes
+      if (vote.proposal_id != _current_qc.proposal_id)
+         return;
+
+      const hs_proposal_message *p = get_proposal( vote.proposal_id );
+      if (p == nullptr) {
+         if (_errors) ilog(" *** ${id} couldn't find proposal", ("id",_id));
+         if (_errors) ilog(" *** ${id} vote : ${vote}", ("vote", vote)("id",_id));
+         return;
+      }
+
+      bool quorum_met = _current_qc.quorum_met; //check if quorum already met
+
+      // If quorum is already met, we don't need to do anything else. Otherwise, we aggregate the signature.
+      if (!quorum_met){
+
+         std::unique_lock state_lock( _state_mutex );
+         if (_current_qc.active_finalizers>0)
+            _current_qc.active_agg_sig = fc::crypto::blslib::aggregate({_current_qc.active_agg_sig, vote.sig });
+         else
+            _current_qc.active_agg_sig = vote.sig;
+
+         _current_qc.active_finalizers = update_bitset(_current_qc.active_finalizers, vote.finalizer);
+         state_lock.unlock();
+
+         quorum_met = is_quorum_met(_current_qc, _schedule, *p);
+
+         if (quorum_met){
+
+            if (_log) ilog(" === ${id} quorum met on #${block_num} ${phase_counter} ${proposal_id} ",
+                           ("block_num", p->block_num())
+                           ("phase_counter", p->phase_counter)
+                           ("proposal_id", vote.proposal_id)
+                           ("id", _id));
+
+            state_lock.lock();
+            _current_qc.quorum_met = true;
+            state_lock.unlock();
+
+            //ilog(" === update_high_qc : _current_qc ===");
+            update_high_qc(_current_qc);
+
+            //check for leader change
+            leader_rotation_check();
+
+            //if we're operating in event-driven mode and the proposal hasn't reached the decide phase yet
+            if (_chained_mode == false && p->phase_counter < 3) {
```

Comment:
`_chained_mode` is only `false` in existing code. I thought we were using chained mode. Is this something different? @systemzax ?

Reply:
Currently we support two hotstuff algorithms : chained and event-driven, with event-driven being the default. Chained Hotstuff does not attempt to make finality progress between blocks. Under chained hotstuff, the leader does not generate a new proposal upon reaching quorum, but rather waits for the next block. This implies that the phase counter for these proposals is always equal to 0 under chained mode.  There are probably only a few scenarios were chained hotstuff is preferrable to event-driven hotstuff ; it does produce slightly lower overhead, since it requires fewer proposals and votes messages, and it might make things a bit easier to debug or to understand. Considering we only need to set a flag to switch from one mode to the other, I figured we might as well keep it as an option.